backend/dataall/cdkproxy/stacks/environment.py

```diff
@@ -775,22 +718,63 @@ def zip_code(assetspath, s3_key='profiler'):
         shutil.make_archive(base_name=f'{assetspath}/{s3_key}', format='zip', root_dir=f'{assetspath}')
         return assetspath
 
-    def set_dlq(self, queue_name) -> sqs.Queue:
-        queue_key = kms.Key(
+    def set_cr_kms_key(self, group_roles) -> kms.Key:
+        key_policy = iam.PolicyDocument(
+            assign_sids=True,
+            statements=[
+                iam.PolicyStatement(
+                    actions=[
+                        "kms:Encrypt",
+                        "kms:Decrypt",
+                        "kms:ReEncrypt*",
+                        "kms:GenerateDataKey*",
+                    ],
+                    effect=iam.Effect.ALLOW,
+                    principals=[
+                        self.pivot_role,
+                        self.default_role,
+                    ] + group_roles,
+                    resources=["*"],
+                    conditions={
+                        "StringEquals": {"kms:ViaService": f"sqs.{self._environment.region}.amazonaws.com"}
+                    }
+                ),
+                iam.PolicyStatement(
+                    actions=[
+                        "kms:DescribeKey",
+                        "kms:List*",
+                        "kms:GetKeyPolicy",
+                    ],
+                    effect=iam.Effect.ALLOW,
+                    principals=[
+                        self.default_role,
+                    ] + group_roles,
+                    resources=["*"],
+                )
+            ]
```

Comment:
Just for clarification. Here we are creating the key for the custom resources DLQ. Since this is an SQS queue, only via SQS the messages should get encrypted and decrypted. The CRs use the pivot role right? so it needs to use the key, but in addition we add the groups to the policy so that they can check the key

Reply:
Yes that is exactly correct - otherwise the `onFailure` messages sent from Lambda to SQS (via the PivotRole) would not be able to be polled off the queue by the environment user. Meaning if a envUser A creates a dataset and the CR Fails to add LF Permissions and the stack fails - envUserA is able to debug by polling message off of the Queue. This is restricted to Encrypting/Decrypting only via SQS!

Also I gave the envUsers the ability to Describe the Key and only GET the Key Policy so these users can understand the permissions they have for this key